src/go/configgenerator/route_generator.go

```diff
@@ -83,12 +84,79 @@ func MakeRouteConfig(serviceInfo *configinfo.ServiceInfo) (*routepb.RouteConfigu
 	host.Routes = append(host.Routes, makeCatchAllNotFoundRoute())
 
 	virtualHosts = append(virtualHosts, &host)
+
+	requestHeaders, err := makeRequestHeadersToAdd(serviceInfo)
+	if err != nil {
+		return nil, err
+	}
+	responseHeaders, err := makeResponseHeadersToAdd(serviceInfo)
+	if err != nil {
+		return nil, err
+	}
 	return &routepb.RouteConfiguration{
-		Name:         routeName,
-		VirtualHosts: virtualHosts,
+		Name:                 routeName,
+		VirtualHosts:         virtualHosts,
+		RequestHeadersToAdd:  requestHeaders,
+		ResponseHeadersToAdd: responseHeaders,
 	}, nil
 }
 
+func makeHeaders(headers string, a bool) ([]*corepb.HeaderValueOption, error) {
+	var l []*corepb.HeaderValueOption
+	for _, h := range strings.Split(headers, ";") {
+		if h == "" {
+			continue
+		}
+		key_value := strings.Split(h, "=")
+		if len(key_value) != 2 {
+			return l, fmt.Errorf("invalid header: %v. should be in key=value format.", h)
+		}
+		if key_value[0] == "" {
+			return l, fmt.Errorf("header key should not be empty for: %v.", h)
+		}
+		l = append(l, &corepb.HeaderValueOption{
+			Header: &corepb.HeaderValue{
+				Key:   key_value[0],
```

Comment:
Should we have any validation for duplicate keys? What will Envoy do in these two cases?

```
addRequestHeaders: k=a;k=b
```
```
addRequestHeaders: k=a
appendRequestHeaders: k=b
```

If envoy will fail startup, no need for validation on our side. If envoy will randomly honor only one of the rules, I would prefer we validate against it.

Reply:
Check Envoy code, they don't verify.